### DISCUSSION TITLE
TrayIcon, how to show NativeMenu with left mouse button single click ?

### DISCUSSION THREAD

**Discussion by xiaokun** [AvaloniaUI/Avalonia#15859] [Q&A]
The default behaviour of TrayIcon to show NativeMenu is right click on it, but right click is more difficult than left click on touchable screen.

``` 
 <TrayIcon.Icons>
      <TrayIcons>
          <TrayIcon Icon="/Assets/logo.ico" Clicked="TrayIcon_Clicked">
              <TrayIcon.Menu>
                  <NativeMenu>
                      <NativeMenuItem Click="OnNetSettingsClicked" Header="Net..."/>
				<NativeMenuItem Click="OnMaskSettingsClicked" Header="Mask..." />
				<NativeMenuItemSeparator />
                      <NativeMenuItem Click="OnExitClicked" Header="Exit..." />
                  </NativeMenu>
              </TrayIcon.Menu>
          </TrayIcon>
      </TrayIcons>
  </TrayIcon.Icons>
```

We can add Clicked event to TrayIcon, but i found no way to open NatvieMenu programmatically in it. 

Does anyone know of any other solution?

**Comment by maxkatz6** [comment]
Left click currently executes Clicked event.
It should be possible to also handle touch message to open the menu, but not clear what to do with Clicked event then.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/TrayIconImpl.cs

**Comment by Green7** [comment]
@maxkatz6 This should be resolved better way. Perhaps, for example, the TrayIcon could have a ShowMenu function?

Currently, on Windows, to display the menu in response to a left-click, I had to search the application's window list and send WM_USER message to the window handling the tray. Which is not an elegant solution ...

**Reply by maxkatz6** [reply]
TryShowMenu API is better than nothing, I suppose. Non-windows platforms should be considered.

**Comment by brandtks** [comment]
Here's a Window's only solution I found to show the menu on a left click. 

``` C#
private void OnTrayIconClicked (object? sender, EventArgs e) {
	// Get this tray icon's implementation
	ITrayIconImpl? impl = (ITrayIconImpl?)GetType ()
		.GetProperty ("Impl", BindingFlags.NonPublic | BindingFlags.Instance)?
		.GetValue (this);

	// Get the Windows tray icon implementation type
	Type? type = AppDomain.CurrentDomain.GetAssemblies ()
		.Where (a => a.FullName?.StartsWith ("Avalonia.Win32") ?? false)
		.SelectMany (a => a.GetTypes ())
		.FirstOrDefault (t => t.Name == "TrayIconImpl");

	// If the Implementation and type are not null
	if (impl != null && type != null) {
		// Get the OnRightClicked method
		MethodInfo? methodInfo = type.GetMethod ("OnRightClicked",
			BindingFlags.NonPublic | BindingFlags.Instance);
		// Invoke the method on the implementation
		methodInfo?.Invoke (impl, null);
	}
}
````

**Reply by Green7** [reply]
This solution works, and it's simpler than my method based on WM_USER.
